sp/src/game/client/mapbase/vscript_vgui.cpp

```diff
@@ -0,0 +1,1871 @@
+// Exposing a new panel class (e.g. vgui::FileOpenDialog):
+//
+// Create 'class IScript_FileOpenDialog : IScript_Frame' with C++ bindings.
+// Create 'class CScript_FileOpenDialog : FileOpenDialog' with vgui message callbacks and overrides (if needed)
+// Create 'class CScript_FileOpenDialog_Helper : IScript_FileOpenDialog< CScript_FileOpenDialog >' for the script bindings
+// Add new condition in CScriptVGUI::Create
+//
+//
+
+
+using namespace vgui;
+class IScriptVGUIObject;
+class CScriptVGUIPanel;
+struct FontData_t;
+
+
+CUtlDict< FontData_t > g_ScriptFonts( k_eDictCompareTypeCaseSensitive );
+CUtlVector< int > g_ScriptTextureIDs;
+CUtlVector< IScriptVGUIObject* > g_ScriptPanels;
+
+
+// Boundary is not checked in Surface, keep count manually to sanitise user input.
+static int g_nFontCount = 0;
+
+static inline unsigned int IntToFontHandle( int i )
+{
+	if ( i < 0 || i > g_nFontCount )
+		return INVALID_FONT;
+	return static_cast< unsigned int >(i);
+}
+
+static inline int HandleToInt( unsigned int i )
+{
+	Assert( i <= INT_MAX );
+	return static_cast< int >(i);
+}
+
+
+struct FontData_t
+{
+	HFont font;
+	char *name;
+	int tall;
+	int weight;
+	int blur;
+	int scanlines;
+	int flags;
+	//int rangeMin;
+	//int rangeMax;
+	bool proportional;
+};
+
+static const char *GetFixedFontName( const char *name, bool proportional )
+{
+	static char fontName[64];
+	V_snprintf( fontName, sizeof(fontName), "%s-%s", name, proportional ? "p" : "no" );
+	return fontName;
+}
+
+CON_COMMAND( vgui_spew_fonts_script, "" )
+{
+	char fontName[64];
+
+	FOR_EACH_DICT_FAST( g_ScriptFonts, i )
+	{
+		const FontData_t &data = g_ScriptFonts[i];
+		const char *name = surface()->GetFontName( data.font );
+		const char *alias = g_ScriptFonts.GetElementName(i);
+
+		// Strip off the appendix "-p" / "-no"
+		V_StrLeft( alias, V_strlen(alias) - (data.proportional ? 2 : 3), fontName, sizeof(fontName) );
+
+		Msg( "  %2d: HFont:0x%8.8x, %s, %s, font:%s, tall:%d(%d)\n",
+			i,
+			data.font,
+			fontName,
+			alias,
+			name ? name : "??",
+			surface()->GetFontTall( data.font ),
+			surface()->GetFontTallRequested( data.font ) );
+	}
+}
+
+void ReloadFontGlyphs()
+{
+	FOR_EACH_DICT_FAST( g_ScriptFonts, i )
+	{
+		const FontData_t &data = g_ScriptFonts[i];
+
+		int tall = data.tall;
+		int blur = data.blur;
+		int scanlines = data.scanlines;
+
+		if ( data.proportional )
+		{
+			tall = scheme()->GetProportionalScaledValue( tall );
+			blur = scheme()->GetProportionalScaledValue( blur );
+			scanlines = scheme()->GetProportionalScaledValue( scanlines );
+		}
+
+		surface()->SetFontGlyphSet(
+			data.font,
+			data.name,
+			tall,
+			data.weight,
+			blur,
+			scanlines,
+			data.flags );
+	}
+}
+
+
+class CScriptRootPanel : public Panel
+{
+private:
+	int m_nLastFrame;
+
+public:
+	CScriptRootPanel()
+	{
+		Init( 0, 0, 0, 0 );
+		SetVisible( false );
+		SetParent( enginevgui->GetPanel( PANEL_ROOT ) );
+	}
+
+	void OnTick()
+	{
+		if ( m_nLastFrame >= gpGlobals->framecount )
```

Comment:
Why is this check necessary?

Reply:
`OnTick()` is called twice in the same frame before the next.